packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -826,8 +827,15 @@ function haveCompatibleStaticParts(vnode: VStatic, renderer: RendererAPI) {
             }
             const { data } = part;
             const hasMatchingAttrs = validateAttrs(vnode, elm, data, renderer, () => true);
-            const hasMatchingStyleAttr = validateStyleAttr(vnode, elm, data, renderer);
-            const hasMatchingClass = validateClassAttr(vnode, elm, data, renderer);
+            // Explicitly skip hydration validation when static parts don't contain `style` or `className`.
+            // This means the style/class attributes are either static or don't exist on the element and
+            // cannot be effected by hydration.
+            const hasMatchingStyleAttr = shouldValidateAttr(data, 'style')
+                ? validateStyleAttr(vnode, elm, data, renderer)
+                : true;
+            const hasMatchingClass = shouldValidateAttr(data, 'className')
+                ? validateClassAttr(vnode, elm, data, renderer)
+                : true;
```

Comment:
I'm not so sure about this logic. Consider this case: 8a22043f1.

If you have a static element:

```html
<p class="c1" lwc:ref="p">text</p>
```

but then you mutate it in `renderedCallback`:

```js
renderedCallback() {
    this.refs.p.classList.add('hahaha!');
}
```

then we would expect a hydration mismatch error, because the client side is adding the new class (via `renderedCallback`), and `renderedCallback` is not invoked server-side.

This works correctly in `master` (hydration mismatch warning). However, with this PR, the hydration mismatch warning actually gets squelched.

Reply:
That's a good point! 

I hadn't considered that scenario, I'll add a test case for it and look into how to address it.

We'll need to have a way to determine what the static class/style values are during hydration.

Reply:
I think there might be an additional scenario we're skipping here.

If the `renderedCallback` dynamically adds a class to an element without static parts, we won't be able to validate against it either:

```html
<p class="c1">text</p>
```

```js
renderedCallback() {
    this.template.querySelector('p').classList.add('hahahah!');
}
```

Since the `<p>` doesn't have any static parts we're bypassing the hydration validation.

Here's a demo 628e7111f84a8b321777da5e95be6b046b569c81 from `master`.

The classes don't match but there's no hydration validation error.

Reply:
During this function, we are comparing two things:
1. the element created as a result of SSR HTML
2. the VDOM generated client-side

`renderedCallback` is not involved in creating thing 1. It is also not involved in creating thing 2; `renderedCallback` may mutate the DOM directly, but it has no mechanism to mutate the VDOM. Additionally, at the point in time when this code is running, `renderedCallback` has not executed yet. It operates after VDOM is generated & the diffing algo updates corresponding elements.

In other words, this scenario should not be a concern & I believe the logic in the PR should be fine.

Reply:
That said, I don't know why it results in a hydration mismatch in `main`... that's a little concerning. Let me take a quick look at that.

Reply:
Looking at https://github.com/salesforce/lwc/commit/8a22043f1c176b305779b950ff911f335a5eece3, I don't know why that would fail in `main`. I'm kind of thinking that may have been a failure case that _shouldn't_ have been a failure case, i.e. it should pass hydration without issue.

Reply:
I was wondering if this was something that changed due to native custom element lifecycle, but I just tested with `DISABLE_NATIVE_CUSTOM_ELEMENT_LIFECYCLE = true` and I'm still getting the same result (hydration mismatch) from 8a22043.

I think you might be right @divmain – my test case is a little busted because the ordering goes like this:

- `hydrateComponent`
- `renderedCallback`
- `test()` <-- checks that the `p.className` is what we expect

I guess if we only care about validating SSR'd HTML against the VDOM at the moment _before_ `connectedCallback`/`renderedCallback`, then yes the current code should be fine.

Reply:
> Looking at 8a22043f1c176b305779b950ff911f335a5eece3, I don't know why that would fail in `main`.

It fails because it needs the fix from this PR. 🙂 If you apply this PR on top of that commit, then the only error is from the `test()` function itself (because it runs after `renderedCallback` and checks that the `p.classList` is what it expects). There are no hydration mismatches.